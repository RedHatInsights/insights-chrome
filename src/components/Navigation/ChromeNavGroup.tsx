import React from 'react';
import PropTypes from 'prop-types';
import { NavGroup } from '@patternfly/react-core';

import WrenchIcon from '@patternfly/react-icons/dist/js/icons/wrench-icon';
import SecurityIcon from '@patternfly/react-icons/dist/js/icons/security-icon';
import TrendUpIcon from '@patternfly/react-icons/dist/js/icons/trend-up-icon';
import CodeIcon from '@patternfly/react-icons/dist/js/icons/code-icon';
import DatabaseIcon from '@patternfly/react-icons/dist/js/icons/database-icon';
import CloudIcon from '@patternfly/react-icons/dist/js/icons/cloud-upload-alt-icon';

import ChromeNavItemFactory from './ChromeNavItemFactory';
import { NavItem } from '../../@types/types';

const sectionTitleMapper = {
  wrench: <WrenchIcon />,
  shield: <SecurityIcon />,
  database: <DatabaseIcon />,
  cloud: <CloudIcon />,
  code: <CodeIcon />,
  'trend-up': <TrendUpIcon />,
};

export type ChromeNavGroupProps = {
  navItems: NavItem[];
  isHidden?: boolean;
  icon?: keyof typeof sectionTitleMapper;
  title: string;
};

const ChromeNavGroup = ({ navItems, isHidden, icon, title }: ChromeNavGroupProps) => {
  const filteredFedrampNavItems = navItems;

  if (isHidden) {
    return null;
  }

  const groupTitle = (
    <div>
      {icon && sectionTitleMapper[icon]}
      {title}
    </div>
  );
  return (
    // PF does not allow node/element in the NavGroup prop types, hence the type cast
    <NavGroup className="chr-c-section-nav" id={title} title={groupTitle as unknown as string}>
      {filteredFedrampNavItems.map((item, index) => (
        <ChromeNavItemFactory key={index} {...item} />
      ))}
    </NavGroup>
  );
};

ChromeNavGroup.propTypes = {
  navItems: PropTypes.array.isRequired,
  icon: PropTypes.oneOf(['wrench', 'shield', 'trend-up', 'database', 'cloud', 'code']),
  title: PropTypes.string.isRequired,
  isHidden: PropTypes.bool,
};

export default ChromeNavGroup;
