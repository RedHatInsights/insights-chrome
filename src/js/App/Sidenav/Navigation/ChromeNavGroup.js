import React from 'react';
import PropTypes from 'prop-types';
import { NavGroup } from '@patternfly/react-core';

import WrenchIcon from '@patternfly/react-icons/dist/js/icons/wrench-icon';
import SecurityIcon from '@patternfly/react-icons/dist/js/icons/security-icon';
import TrendUpIcon from '@patternfly/react-icons/dist/js/icons/trend-up-icon';
import ChromeNavItemFactory from './ChromeNavItemFactory';

const sectionTitleMapper = {
  wrench: <WrenchIcon />,
  shield: <SecurityIcon />,
  'trend-up': <TrendUpIcon />,
};

const ChromeNavGroup = ({ navItems, isHidden, icon, title }) => {
  if (isHidden) {
    return null;
  }

  const groupTitle = (
    <div>
      {icon && sectionTitleMapper[icon]}
      {title}
    </div>
  );
  return (
    <NavGroup className="ins-c-section-nav" id={title} title={groupTitle}>
      {navItems.map((item, index) => (
        <ChromeNavItemFactory key={index} {...item} />
      ))}
    </NavGroup>
  );
};

ChromeNavGroup.propTypes = {
  navItems: PropTypes.array.isRequired,
  icon: PropTypes.oneOf(['wrench', 'shield', 'trend-up']),
  title: PropTypes.string.isRequired,
  isHidden: PropTypes.bool,
};

export default ChromeNavGroup;
