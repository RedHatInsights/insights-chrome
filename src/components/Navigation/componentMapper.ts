import ChromeNavGroup from './ChromeNavGroup';
import ChromeNavExapandable from './ChromeNavExapandable';
import ChromeNavItem from './ChromeNavItem';
import DynamicNav from './DynamicNav';

const componentMapper = {
  group: ChromeNavGroup,
  expandable: ChromeNavExapandable,
  item: ChromeNavItem,
  dynamicNav: DynamicNav,
};

export default componentMapper;
