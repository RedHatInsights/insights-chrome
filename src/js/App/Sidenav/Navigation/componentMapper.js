import ChromeNavGroup from './ChromeNavGroup';
import ChromeNavExapandable from './ChromeNavExapandable';
import ChromeNavItem from './ChromeNavItem';

const componentMapper = {
  group: ChromeNavGroup,
  expandable: ChromeNavExapandable,
  item: ChromeNavItem,
};

export default componentMapper;
