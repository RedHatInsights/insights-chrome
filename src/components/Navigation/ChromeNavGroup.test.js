/* eslint-disable react/prop-types */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import createMockStore from 'redux-mock-store';
import NavContext from './navContext';
import ChromeNavGroup from './ChromeNavGroup';
import componentMapper from './componentMapper';

const NavContextWrapper = ({
  store,
  providerValue = {
    onLinkClick: jest.fn(),
    componentMapper,
  },
  children,
  initialEntries,
}) => (
  <MemoryRouter initialEntries={initialEntries}>
    <Provider store={store}>
      <NavContext.Provider value={providerValue}>{children}</NavContext.Provider>
    </Provider>
  </MemoryRouter>
);

describe('ChromeNavGroup', () => {
  const mockStore = createMockStore();
  const groupTitle = 'Foo';
  const testProps = {
    appId: 'testModule',
    href: '/foo',
    title: groupTitle,
    navItems: [],
  };
  const store = mockStore({
    chrome: {
      activeModule: 'testModule',
      moduleRoutes: [],
      modules: {
        testModule: {},
      },
    },
  });

  test('should not render nav item group', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavGroup isHidden {...testProps} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(groupTitle)).toHaveLength(0);
  });

  test('should render nav item group', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavGroup {...testProps} />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(groupTitle)).toHaveLength(1);
  });

  test('should render nav item group with icon', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavGroup {...testProps} icon="wrench" />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(groupTitle)).toHaveLength(1);
  });

  test('should render nav item group with items', () => {
    const { queryAllByText, container } = render(
      <NavContextWrapper store={store}>
        <ChromeNavGroup
          {...testProps}
          navItems={[
            {
              appId: 'testModule',
              href: '/foo',
              title: 'Sub Item',
            },
          ]}
        />
      </NavContextWrapper>
    );
    expect(container).toMatchSnapshot();
    expect(queryAllByText(groupTitle)).toHaveLength(1);
    expect(queryAllByText('Sub Item')).toHaveLength(1);
  });
});
