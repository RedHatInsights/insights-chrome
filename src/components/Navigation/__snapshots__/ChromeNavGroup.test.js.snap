// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromeNavGroup should not render nav item group 1`] = `<div />`;

exports[`ChromeNavGroup should render nav item group 1`] = `
<div>
  <section
    aria-labelledby="Foo"
    class="pf-c-nav__section chr-c-section-nav"
  >
    <h2
      class="pf-c-nav__section-title"
      id="Foo"
    >
      <div>
        Foo
      </div>
    </h2>
    <ul
      class="pf-c-nav__list chr-c-section-nav"
    />
  </section>
</div>
`;

exports[`ChromeNavGroup should render nav item group with icon 1`] = `
<div>
  <section
    aria-labelledby="Foo"
    class="pf-c-nav__section chr-c-section-nav"
  >
    <h2
      class="pf-c-nav__section-title"
      id="Foo"
    >
      <div>
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 512 512"
          width="1em"
        >
          <path
            d="M507.73 109.1c-2.24-9.03-13.54-12.09-20.12-5.51l-74.36 74.36-67.88-11.31-11.31-67.88 74.36-74.36c6.62-6.62 3.43-17.9-5.66-20.16-47.38-11.74-99.55.91-136.58 37.93-39.64 39.64-50.55 97.1-34.05 147.2L18.74 402.76c-24.99 24.99-24.99 65.51 0 90.5 24.99 24.99 65.51 24.99 90.5 0l213.21-213.21c50.12 16.71 107.47 5.68 147.37-34.22 37.07-37.07 49.7-89.32 37.91-136.73zM64 472c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
          />
        </svg>
        Foo
      </div>
    </h2>
    <ul
      class="pf-c-nav__list chr-c-section-nav"
    />
  </section>
</div>
`;

exports[`ChromeNavGroup should render nav item group with items 1`] = `
<div>
  <section
    aria-labelledby="Foo"
    class="pf-c-nav__section chr-c-section-nav"
  >
    <h2
      class="pf-c-nav__section-title"
      id="Foo"
    >
      <div>
        Foo
      </div>
    </h2>
    <ul
      class="pf-c-nav__list chr-c-section-nav"
    >
      <li
        class="pf-c-nav__item"
        data-ouia-component-id="Sub Item"
        data-ouia-component-type="PF4/NavItem"
        data-ouia-safe="true"
      >
        <a
          class="pf-c-nav__link"
          data-quickstart-id="foo"
          data-testid="router-link"
          href="/foo"
        >
          Sub Item
           
        </a>
      </li>
    </ul>
  </section>
</div>
`;
